src/atomic.rs

```diff
@@ -0,0 +1,860 @@
+use std::borrow::{Borrow, BorrowMut};
+use std::marker::PhantomData;
+use std::mem;
+use std::ops::{Deref, DerefMut};
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+
+use Scope;
+
+/// Given ordering for the success case in a compare-exchange operation, returns the strongest
+/// appropriate ordering for the failure case.
+#[inline]
+fn strongest_failure_ordering(ord: Ordering) -> Ordering {
```

Comment:
Fair enough, let's go with `compare_exchange` then. :)
I've renamed `CasOrdering` to `CompareExchangeOrdering`. It's a bit lengthy, but the meaning is clear.

About returning `Ok(previous)`... My main gripe is with the return type of `compare_exchange_owned`. At the moment it returns:
```rust
Result<Ptr<'scope, T>, (Ptr<'scope, T>, Owned<T>)>
```

If we change the signature so that it returns the previous value in the success case as well, the return type would have to be:
```rust
Result<(Ptr<'scope, T>, Ptr<'scope, T>), (Ptr<'scope, T>, Owned<T>)>
```
This is a bit long and possibly error-prone because nothing distinguishes the two pointers in the success case (the new and the previous pointer).

For that reason, I think it's best not to return the previous value in the success case.
But now the danger is that someone might take a look at these signatures and think that they follow the same conventions:
```rust
AtomicPtr::compare_exchange(...) -> Result<*mut T, *mut T>
Atomic::compare_exchange_owned(...) -> Result<Ptr<'scope, T>, (Ptr<'scope, T>, Owned<T>)>
```

But they don't! :) The first function returns the *previous* value in the success case, while the second returns the *new* value. I'm afraid that this might cause some confusion, but perhaps I'm wrong.